pkg/storage/storagewrappers/sharediterator/shared_iterator_datastore.go

```diff
@@ -135,36 +148,55 @@ func (sf *IteratorDatastore) ReadStartingWithUser(
 
 	if found {
 		keyItem.counter++
+		keyItem.iter.mu.Lock()
+		defer keyItem.iter.mu.Unlock()
```

Comment:
> sf.internalStorage.mu.Unlock()

Should be the first line in the `if found {` block. Marginal, but if the same `keyItem` is under contention, then the whole thing would be slowed down.

Reply:
I will move this up to be before keyItem.iter.mu.Lock().  However, this needs to be after keyItem.counter++ so that we won't prematurely delete the key.